enhancements/machine-api/control-plane.md

```diff
@@ -0,0 +1,339 @@
+
+#### Story 4
+- As a SRE, I want to have consumable API primitives in place to resize Control Plane compute resources so I can develop upper level automation tooling atop. E.g Control Plane autoresizing to support a severe growth peak of the number of worker nodes.
+
+#### Story 5
+- As an operator, I want my machine API cluster infrastructure to be self healing so I want faulty nodes to be remediated automatically for safety scenarios.
+
+#### Story 6
+- As a multi cluster operator, I want to have a universal user experience for managing the Control Plane in a declarative manner across any cloud provider, bare metal and any flavour of the product that have in common the topology assumed in this doc.
+
+#### Story 7
+- As a developer, I want to be able to deploy the smallest possible cluster to save costs, i.e one instance Control Plane.
+
+### Implementation Details/Notes/Constraints [optional]
+
+To satisfy the goals, motivation and stories above this proposes a new CRD and a controller that manages the Control Plane as a single entity. It supports the following features:
+
+#### Declarative horizontal scaling
+##### Scale out
+1. The controller always reconciles towards expected number of replicas. Validation enforce this to be an odd number.
+2. Fetch all existing control plane Machine resources by ownerRef. Adopt any other machine having a targeted label e.g `node-role.kubernetes.io/master`.
+3. Compare with expected replicas number. If expected is higher than current then:
+4. Check all owned machines have a backed ready node.
+5. Check all etcd members for all owned machines are healthy via Cluster etcd Operator status signalling.
+6. If (NOT all etcd members are healthy OR NOT all owned machines have a backed ready node) then controller short circuits here, log, update status and requeue. Else:
+7. Choose a failure domain.
+8. Create new Machine object with a templated spec. Go to 1.
+9. Cluster etcd Operator watches the new node. It runs a new etcd pod on it.
+
+##### Scale in
+1. The controller always reconciles towards expected number of replicas. Validation enforce this to be an odd number.
+2. Fetch all existing control plane Machine resources by ownerRef. Adopt any other machine having a targeted label e.g `node-role.kubernetes.io/master`.
+3. Compare with expected replicas number. If expected is lower than current then:
+4. Check all owned machines have a backed ready node.
+5. Check all etcd members for all owned machines are healthy via Cluster etcd Operator status signalling.
+6. If (NOT all etcd members are healthy OR NOT all owned machines have a backed ready node) then controller short circuits here, log, update status and requeue. Else:
+7. Pick oldest machine in more populated failure domain out of a candidates to be deleted list (by default all owned Machines).
+8. Remove etcd member.
```

Comment:
isnt there a 'remove-node' api? So the remove etcd-member will be done by the CEO rather than this controller?

Reply:
scaling etcd is owned by cluster-etcd-operator.